File: Virucodesoup/Ruby/OFB (Output Feedback)/ofb.rb
# ===========================
# File : ofb.rb
#
# Credit:
# - Coded by (aka in cyber) Vsio Stitched
#
# Misc:
# - Written in Ruby programming language
#
# License:
# - Free to use
# - May include me or not in credit if included in other project
# ===========================

class OFB
	attr_accessor:inputText_,:key_,:outputText_
	# inputText_: text input whicih will be encrypted or decrypted
	# key_: binary input
	# outputText_: text output which is resulted from encryption or decryption
	
	def initialize()
	# inits variables value since initialization
		
		@inputText_  = ""
		@key_ = ""
		@outputText_ = ""
	end
	
	def menu()
	# displays command menu
	
		index = 0
		while (index!="3")
			puts "\n\n==> MENU\n\n"
			puts "1. Encrypt"
			puts "2. Decrypt"
			puts "3. Exit"
			print "\nSelect a number : ";
			index = gets.strip
			puts "\n"
			
			case index
			
				when "1" then
					loadFile()
					inputKey()
					encipher()
					saveFile(0)
					break

				when "2" then
					loadFile()
					inputKey()
					decipher()
					saveFile(1)
					break
					
				when "3" then 
					break
					
				else 
					puts "> Invalid Input"
					index = 0;
			end
		end
	end
	
	def inputKey()
	# input key_ value
	
		value = " "
		regexConstraint = /[^0-1]/
		while (regexConstraint.match(value))
			print "\nInput binary string: "
			value = gets.strip
			if (regexConstraint.match(value))
				puts "> Error: Only 0 or 1 is allowed"					
			end
		end
		
		@key_ = value
	end
	
	def generateIV()
	# generates Initialization Vector randomly
	
		blockLength = @key_.length
		iv = ""
		
		for i in 0..blockLength-1
			iv += rand.round.to_s
		end
		
		print "\n> IV generated: ", iv, "\n"
		
		saveIV(iv)
		
		return iv
	end
	
	def encryptionBlock(_plainText)
	# encryption blocks (in this example, a simple xor encryption is used) (You can define your own Encryption Block here)
		
		output = ""
		key = @key_
		for i in 0..key.length-1
			output += (_plainText[i].to_i ^ key[i].to_i).to_s			
		end
		
		return output
	end
	
	def decryptionBlock(_cipherText)
	# decryption blocks (in this example, a simple xor decryption is used) (You can define your own Decryption Block here)
		
		output = ""
		key = @key_
		for i in 0..key.length-1
			output += (_cipherText[i].to_i ^ key[i].to_i).to_s			
		end
		
		return output
	end
	
	def encipher()
	# enciphers with OFB method
	
		inputBit = @inputText_.unpack("B*")[0]
		inputBitLength = inputBit.length		
		blockLength = @key_.length
		iv = generateIV()
		blockTemp = ""
		blockCipher = ""
		output = ""
		outputText = ""
		i = 0
		j = 0
		
		blockCipher = encryptionBlock(iv);
		
		while (inputBit.length % blockLength != 0)
			inputBit += '0'
		end
		
		for i in 0..blockLength-1
			blockTemp += (blockCipher[i].to_i ^ inputBit[j].to_i).to_s			
			j += 1			
		end
		
		output += blockTemp
		blockCipher = encryptionBlock(blockCipher);
		blockTemp = ""
		
		while (j != inputBit.length)			
			for i in 0..blockLength-1				
				blockTemp += (blockCipher[i].to_i ^ inputBit[j].to_i).to_s
				
				j += 1
			end	
			output += blockTemp
			blockCipher = encryptionBlock(blockCipher);
			blockTemp = ""
		end
		
		output = output[0..inputBitLength-1]
		
		i = 0
		
		while (i!=output.length)
			outputText += [output[i..i+7]].pack("B*")
			i+=8
		end
		
		@outputText_ = outputText
		puts "\n>Encryption Result:"
		puts outputText
	end
	
	def decipher()
	# deciphers with OFB method

		inputBit = @inputText_.unpack("B*")[0]
		inputBitLength = inputBit.length
		blockLength = @key_.length
		blockPlain = ""
		blockCipher = ""
		blockTemp = ""
		iv = loadIV()
		output = ""
		outputText = ""
		i = 0
		j = 0
		
		blockCipher = decryptionBlock(iv);
		
		while (inputBit.length % blockLength != 0)
			inputBit += '0'
		end
		
		for i in 0..blockLength-1
			blockPlain += (blockCipher[i].to_i ^ inputBit[j].to_i).to_s
			j += 1			
		end
		
		output += blockPlain
		blockCipher = decryptionBlock(blockCipher);
		blockPlain = ""
		
		while (j != inputBit.length)			
			for i in 0..blockLength-1				
				blockPlain += (inputBit[j].to_i ^ blockCipher[i].to_i).to_s			
				j += 1
			end	
						
			output += blockPlain
			blockCipher = decryptionBlock(blockCipher);
			blockPlain = ""
		end
		
		output = output[0..inputBitLength-1]
		
		i = 0
		
		while (i!=output.length)
			outputText += [output[i..i+7]].pack("B*")
			i+=8
		end
		
		@outputText_ = outputText
		puts "\n>Decryption Result:"
		puts outputText
		
	end
	
	def loadFile()
	# loads input from a file
		
		begin
			print "Input File: "
			fileString = gets.strip
			file = File.new(fileString,"rb")			
			buffer = ''
			
			file.each_line do |line|
				buffer += line
			end
			
			@inputText_ = buffer
			puts "\n"+@inputText_
			file.close
		rescue
			puts "> File not found. Try again."
			loadFile()
		end
	end

	def loadIV()
	# loads Initialization Vector from a file
		
		begin
			print "Input IV File: "
			fileString = gets.strip
			file = File.new(fileString,"rb")			
			buffer = ''
			
			file.each_line do |line|
				buffer += line
			end
			
			output = buffer
			puts "\n"+output
			file.close
		rescue
			puts "> File not found. Try again."
			loadFile()
		end
		
		return output
	end
	
	def saveFile(_index)
	# saves output a file
	
		if (_index == 0)
			begin
				file = File.new("output_encipher.txt","wb")			
				file.print @outputText_;
				file.close;
				puts "\n> OFB enciphering result has been saved into output_encipher.txt"
			rescue
				puts "> File not found."
			end
		elsif (_index == 1)
			begin
				file = File.new("output_decipher.txt","wb")			
				file.print @outputText_;
				file.close;
				puts "\n> OFB deciphering result has been saved into output_decipher.txt"
			rescue
				puts "> File not found."
			end
		end
	end	
	
	def saveIV(_iv)
	# saves output a file
	
		begin
			file = File.new("iv.txt","wb")			
			file.print _iv;
			file.close;
			puts "\n> IV has been saved into iv.txt"
		rescue
			puts "> File not found."
		end
	end	
	
	def driver()
	# simulates instance functionality
		
		menu()
	end
	
end

File: Virucodesoup/C#/Standard/OFB (Output Feedback)/ofb.cs
/* ===========================
File : ofb.cs

Credit:
- Coded by (aka in cyber) Vsio Stitched
- Problem case inspired from my past in high school

Misc:
- Written in C# programming laguage

License:
- Free to use
- May include me or not in credit if included in other project
=========================== */

using System;
using System.IO;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;


class OFB {

    // FIELD

    private String inputText_; /* text input whicih will be encrypted or decrypted */
    private String key_; /* binary input */
    private String outputText_; /* text output which is resulted from encryption or decryption */
	private byte[] outputBytes_; /* output in bytes */

	
    // CONSTRUCTOR

    public OFB()
    /* constructs OFB instance */
    {
        this.inputText_ = "";
        this.key_ = "";
        this.outputText_ = "";
    }

    public OFB(String _newInputText, String _newKey, String _newOutputText)
    /* constructs OFB instance */
    {
        this.inputText_ = _newInputText;
        this.key_ = _newKey;
        this.outputText_ = _newOutputText;
    }


    // ACCESSOR GET

    public String getInputText()
    /* gets inputText_ value */
    {
        return this.inputText_;
    }

    public String getKey()
    /* returns key_ value */
    {
        return this.key_;
    }

    public String getOutputText()
    /* returns outputText_ value */
    {
        return this.outputText_;
    }

	public byte[] getOutputBytes()
	/* returns outputBytes_ value */
	{
		return this.outputBytes_;
	}
	
	
    // ACCESSOR SET

    public void setInputText(String _newInputText)
    /* changes inputText_ value */
    {
        this.inputText_ = _newInputText;
    }

    public void setKey(String _newKey)
    /* changes key_ value */
    {
        this.key_ = _newKey;
    }

    public void setOutputText(String _newOutputText)
    /* changes outputText_ value */
    {
        this.outputText_ = _newOutputText;
    }

	public void setOutputBytes(byte[] _newOutputBytes)
	/* returns outputBytes_ value */
	{
		this.outputBytes_ = _newOutputBytes;
	}
	
	
    // PREDICATE

	
    // METHOD

    public void menu()
    /* displays command menu */
    {
		String index = "0";
                
        while (index[0] != '3') {
            Console.WriteLine("\n\n==> MENU\n\n");
            Console.WriteLine("1. Encrypt");
            Console.WriteLine("2. Decrypt");
            Console.WriteLine("3. Exit");
            Console.Write("\nSelect a number : ");
            index = Console.ReadLine();
            Console.WriteLine();
            switch(index[0]) {
                case '1': {
                    loadFile();
                    inputKey();
                    encipher();
                    saveFile(0);
					goto case '3';
				}
                case '2': {
                    loadFile();
                    inputKey();
                    decipher();
                    saveFile(1);
					goto case '3';
                }
                case '3': {
                    index = "3";
					break;
                }
                default:{
                    Console.WriteLine("> Invalid Input");
                    index = "0";
					break;
                }

            }
        }
    }
	
	public void inputKey()
	/* input key_ value */
	{
		String value = " ";
		Regex regexConstraint = new Regex("^[0-1]+$");
		
		while (!regexConstraint.IsMatch(value)) {
			Console.Write("\nInput binary string: ");
			value = Console.ReadLine();
			if (!regexConstraint.IsMatch(value)) {
				Console.WriteLine("> Error: Only 0 or 1 is allowed");				
			}
		}
		
		this.key_ = value;
	}
	
	public string generateIV()
	/* generates Initialization Vector randomly */
	{
		int blockLength = this.key_.Length;
		string iv = "";
		Random random = new Random();
		
		for (int i=0;i!=blockLength;i++) {
			iv += ((int)(double)random.Next(0,2)).ToString();
		}
		
		Console.Write("\n> IV generated: "+ iv+ "\n");
		
		saveIV(iv);
		
		return iv;
	}
	
	public String encryptionBlock(String _plainText)
	/* encryption blocks (in this example, a simple xor encryption is used) (You can define your own Encryption Block here) */
	{
		String plainText = _plainText;
		String output = "";
		String key = this.key_;
		
		for (int i=0;i!=key.Length;i++) {
			output += xorBit(plainText[i],key[i]);
		}
		
		return output;
	}

	public String decryptionBlock(String _cipherText)
	/* decryption blocks (in this example, a simple xor decryption is used) (You can define your own Decryption Block here) */
	{
		String cipherText = _cipherText;
		String output = "";
		String key = this.key_;
		
		for (int i=0;i!=key.Length;i++) {
			output += xorBit(cipherText[i],key[i]);
		}
		
		return output;
	}
	
	public byte[] binaryStringToBytes(string _binaryString) 
	/* converts binary string to binary bytes */
	{
		string binary = _binaryString;
		int i = 0;
		int j = 0;
		BitArray bits = new BitArray(8);
		byte[] bytes = new byte[binary.Length/8];
		
		while (j!=binary.Length) {		
			for (i=0;i!=8;i++) {
				
				bits.Set(i,(binary[j+i]=='1'?true:false));
			}
			
			bits.CopyTo(bytes,j/8);
			j+=8;
		}
		
		return bytes;
	}
	
	public char binaryToChar(string _binaryString) 
	/* converts binary to char */
	{
		UTF8Encoding utf8 = new UTF8Encoding();
		string binary = _binaryString;
		byte[] bytes = binaryStringToBytes(binary);
		string output = utf8.GetString(bytes);
		
		return output[0];
	}
	
	public string binaryToString(string _binaryString)
	/* converts binary to string */
	{
		string binaries = _binaryString;
		string output = "";
		int i = 0;
		while (i!=binaries.Length) {			
			output += binaryToChar(binaries.Substring(i,8));
			i+=8;
		}
		
		return output;
	}
	
	public char xorBit(char _bit1,char _bit2) 
	/* does xor operator with bits in char types */
	{
		return ((_bit1==_bit2) ? '0':'1') ;
	}
	
	public void encipher()
	/* enciphers with OFB method */
	{	
		string inputBit = this.inputText_;
		int inputBitLength = inputBit.Length;
		int blockLength = this.key_.Length;
		string blockCipher = "";
		string blockTemp = "";
		string iv = generateIV();
		string output = "";
		string outputText = "";
		int i = 0;
		int j = 0;
		
		blockCipher = encryptionBlock(iv);
		
		while (inputBit.Length % blockLength != 0) {
			inputBit += '0';			
		}
		
		for(i=0;i!=blockLength;i++) {
			blockTemp += xorBit(blockCipher[i],inputBit[j]);
			j++;
		}
		
		output += blockTemp;
		blockCipher = encryptionBlock(blockCipher);
		blockTemp = "";
		
		while (j != inputBit.Length) {		
			for (i=0;i!=blockLength;i++) {				
				blockTemp += xorBit(blockCipher[i],inputBit[j]);			
				j += 1;
			}
			output += blockTemp;
			blockCipher = encryptionBlock(blockCipher);
			blockTemp = "";
		}
		
		outputText = binaryToString(output.Substring(0,inputBitLength));	
		this.outputText_ = outputText;		
		this.outputBytes_ = binaryStringToBytes(output.Substring(0,inputBitLength));
		
		Console.WriteLine("\n>Encryption Result:");
		Console.WriteLine(outputText);
	}
	
	public void decipher()
	/* deciphers with OFB method */
	{
		string inputBit = this.inputText_;
		int inputBitLength = inputBit.Length;
		int blockLength = this.key_.Length;
		string blockCipher = "";
		string blockPlain = "";
		string iv = loadIV();
		string output = "";
		string outputText = "";
		int i = 0;
		int j = 0;
		
		blockCipher = decryptionBlock(iv);
		
		while (inputBit.Length % blockLength != 0) {
			inputBit += '0';
		}
		
		for (i=0;i!=blockLength;i++) {
			blockPlain += xorBit(blockCipher[i],inputBit[i]);
			j += 1;
		}

		output += blockPlain;
		blockCipher = decryptionBlock(blockCipher);
		blockPlain = "";
		
		while (j != inputBit.Length) {
			for (i=0;i!=blockLength;i++) {		
				blockPlain += xorBit(inputBit[j],blockCipher[i]);
				j += 1;
			}
			output += blockPlain;
			blockCipher = decryptionBlock(blockCipher);
			blockPlain = "";
		}
		
		outputText = binaryToString(output.Substring(0,inputBitLength));
		this.outputText_ = outputText;
		this.outputBytes_ = binaryStringToBytes(output.Substring(0,inputBitLength));
		
		Console.WriteLine("\n>Decryption Result:");
		Console.WriteLine(outputText);
	}
	
	public void loadFile()
	/* loads input from a file */
	{
		string fileString = "";
		UTF8Encoding utf8 = new UTF8Encoding();
		
		try {			
			Console.Write("\nInput file name: ");
			fileString = Console.ReadLine();
			Console.WriteLine();
			string output = "";			
			byte[] bytes = File.ReadAllBytes(fileString);			
			BitArray bits = new BitArray(bytes);
	
			for (int i=0;i!=bits.Length;i+=1) {
				output += (bits[i] ? '1' : '0');
			}
			
			String inputText = utf8.GetString(bytes);
			Console.WriteLine(inputText);
			this.inputText_ = output;
			
		} catch (Exception) {
			Console.WriteLine("> File not found. Try again.");
			loadFile();
		}
	}
	
	public string loadIV()
	/* loads Initialization Vector from a file */
	{
		StreamReader file;
		string fileString = "";
		string iv = "";
		
		try {			
			Console.Write("\nInput IV File: ");
			fileString = Console.ReadLine();
			Console.WriteLine();
			file = new StreamReader(fileString);	
			iv = file.ReadToEnd();
			Console.WriteLine(iv);
			file.Close();
		} catch (Exception) {
			Console.WriteLine("> File not found. Try again.");
			loadFile();
		}
		
		return iv;
	}
	
	public void saveFile(int _index)
	/* saves output a file */
	{	
		if (_index == 0) {
			try {
				File.WriteAllBytes("output_encipher.txt", this.outputBytes_);				
				Console.WriteLine("\n> OFB enciphering result has been saved into output_encipher.txt");
			} catch (Exception) {
				Console.WriteLine("> File not found.");
			}
		} else if (_index == 1) {
			try {
				File.WriteAllBytes("output_decipher.txt", this.outputBytes_);				
				Console.WriteLine("\n> OFB deciphering result has been saved into output_decipher.txt");
			} catch (Exception) {
				Console.WriteLine("> File not found.");
			}
		}
	}
	
	public void saveIV(string _iv)
	/* saves Initialization Vector into a file */
	{
		try {		
			StreamWriter file = new StreamWriter("iv.txt");
			
			file.AutoFlush = true;
			file.Write(_iv);
				
			Console.WriteLine("\n> IV has been saved into iv.txt");
			file.Close();
		} catch (Exception) {
			Console.WriteLine("> File not found.");
		}
	}
	
    // DRIVER

    public void driver()
    /* simulates instance functionality */
    {
		menu();
    }

    public static void Main(String[] args)
    /* executes class program */
    {
		
        OFB ofb = new OFB();
        
		Console.Write("\n===== OFB (Output Feedback) =====\n");
		Console.Write("\n\n== START PROGRAM ==\n\n");
		ofb.driver();
		Console.Write("\n\n== END PROGRAM ==\n\n");
		Console.ReadLine();
    }
	
}